#### Importing Libraries

In [1]:
import numpy as np # Numpy we are using for numerical calculation in python 
import pandas as pd # pandas we are using for data frame and analysis in python 
import matplotlib.pyplot as plt #matplotlib we are using it for ploting and visuilazing the data in python
import seaborn as sns #seaborn we are using it for ploting and visuilazing the data in python
import statsmodels.api as sm
%matplotlib inline 
sns.set(style="whitegrid")
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score

#### Importing the Housing data

In [2]:
csv_housing_data = pd.read_csv(r"C:\Learning\Data\housing.data")


#### Displaying Top 10 rows of Housing Data

In [3]:
csv_housing_data.head(10)

0.00632  18.00   2.310  0  0.5380  6.5750  65.20  4.0900   1  296.0  15.30 396.90   4.98  24.00
0   0.02731   0.00   7.070  0  0.4690  6.4210  78...                                              
1   0.02729   0.00   7.070  0  0.4690  7.1850  61...                                              
2   0.03237   0.00   2.180  0  0.4580  6.9980  45...                                              
3   0.06905   0.00   2.180  0  0.4580  7.1470  54...                                              
4   0.02985   0.00   2.180  0  0.4580  6.4300  58...                                              
5   0.08829  12.50   7.870  0  0.5240  6.0120  66...                                              
6   0.14455  12.50   7.870  0  0.5240  6.1720  96...                                              
7   0.21124  12.50   7.870  0  0.5240  5.6310 100...                                              
8   0.17004  12.50   7.870  0  0.5240  6.0040  85...                                              
9   0.22489  12.50   7.870  0  0.5240  6.3770  94...

#### Importing the Housing data CSV by considering that there are no headers in the CSV and also considering column separator is whitespace

In [4]:
housing_data = pd.read_csv(r"C:\Learning\Data\housing.data",header=None,delim_whitespace=True)

#### Displaying Top 10 rows of Housing data without Column Headers

In [5]:
housing_data.head(10)

0     1     2   3      4      5      6       7   8      9     10  \
0  0.00632  18.0  2.31   0  0.538  6.575   65.2  4.0900   1  296.0  15.3   
1  0.02731   0.0  7.07   0  0.469  6.421   78.9  4.9671   2  242.0  17.8   
2  0.02729   0.0  7.07   0  0.469  7.185   61.1  4.9671   2  242.0  17.8   
3  0.03237   0.0  2.18   0  0.458  6.998   45.8  6.0622   3  222.0  18.7   
4  0.06905   0.0  2.18   0  0.458  7.147   54.2  6.0622   3  222.0  18.7   
5  0.02985   0.0  2.18   0  0.458  6.430   58.7  6.0622   3  222.0  18.7   
6  0.08829  12.5  7.87   0  0.524  6.012   66.6  5.5605   5  311.0  15.2   
7  0.14455  12.5  7.87   0  0.524  6.172   96.1  5.9505   5  311.0  15.2   
8  0.21124  12.5  7.87   0  0.524  5.631  100.0  6.0821   5  311.0  15.2   
9  0.17004  12.5  7.87   0  0.524  6.004   85.9  6.5921   5  311.0  15.2   

       11     12    13  
0  396.90   4.98  24.0  
1  396.90   9.14  21.6  
2  392.83   4.03  34.7  
3  394.63   2.94  33.4  
4  396.90   5.33  36.2  
5  394.12   5.21  28.7  
6  395.60  12.43  22.9  
7  396.90  19.15  27.1  
8  386.63  29.93  16.5  
9  386.71  17.10  18.9

#### Adding Column names to the data frame

In [6]:
housing_data.columns = ["CRIM","ZN","INDUS","CHAS","NOX","RM","AGE","DIS","RAD","TAX","PTRATIO","B","LSTAT","MEDV"]

#### Displaying Top 5 rows with Header

In [7]:
housing_data.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296.0   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242.0   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242.0   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222.0   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222.0   

   PTRATIO       B  LSTAT  MEDV  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     17.8  392.83   4.03  34.7  
3     18.7  394.63   2.94  33.4  
4     18.7  396.90   5.33  36.2

#### Declaring list of input variables and target/dependant variable

In [8]:
x_columns = ["CRIM","ZN","INDUS","CHAS","NOX","RM","AGE","DIS","RAD","TAX","PTRATIO","B","LSTAT"]
pdx = housing_data[x_columns]
pdy = housing_data["MEDV"]

#### Split the training data and test data by considering 70:30 ratio

In [9]:
x_train, x_test,y_train,y_test = train_test_split(pdx,pdy,train_size=70/100,random_state = 70)

#### Intercepting the data for Ordinary Least Square Model

In [10]:
x_train_cons = sm.add_constant (x_train)
x_test_cons = sm.add_constant (x_test)
full_mod = sm.OLS(y_train,x_train_cons)

#### Fitting the model and Printing the Result

In [11]:
regression_full = full_mod.fit()
print(regression_full.summary())

                            OLS Regression Results                            
Dep. Variable:                   MEDV   R-squared:                       0.738
Model:                            OLS   Adj. R-squared:                  0.728
Method:                 Least Squares   F-statistic:                     73.82
Date:                Mon, 12 Apr 2021   Prob (F-statistic):           1.37e-90
Time:                        07:48:29   Log-Likelihood:                -1038.4
No. Observations:                 354   AIC:                             2105.
Df Residuals:                     340   BIC:                             2159.
Df Model:                          13                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         38.5890      5.969      6.464      0.0

In [12]:
print("Variance Inflation Factor")
cnames = x_train.columns
for i in (np.arange(0,len(cnames))):
    xvars = list(cnames)
    yvar = xvars.pop(i)
    mod = sm.OLS(x_train[yvar],sm.add_constant(x_train_cons[xvars]))
    res = mod.fit()
    vif = 1/(1-res.rsquared)
    print(yvar,round(vif,3))

Variance Inflation Factor
CRIM 1.772
ZN 2.475
INDUS 3.839
CHAS 1.121
NOX 4.592
RM 1.834
AGE 3.162
DIS 4.207
RAD 7.365
TAX 8.645
PTRATIO 1.739
B 1.303
LSTAT 2.769


#### In the above iteration R-Square Value is 0.738 and Adjacent R-Square Value is 0.728. 
#### Probability value is high for the columns AGE - 0.659 & INDUS - 0.451
#### Going to Run one more iteration by removing the independant variable AGE

In [13]:
x_columns = ["CRIM","ZN","INDUS","CHAS","NOX","RM","DIS","RAD","TAX","PTRATIO","B","LSTAT"]
pdx = housing_data[x_columns]
pdy = housing_data["MEDV"]

x_train, x_test,y_train,y_test = train_test_split(pdx,pdy,train_size=70/100,random_state = 70)

x_train_cons = sm.add_constant (x_train)
x_test_cons = sm.add_constant (x_test)
full_mod = sm.OLS(y_train,x_train_cons)

regression_full = full_mod.fit()
print("Iteration-2", regression_full.summary())

print("Variance Inflation Factor")
cnames = x_train.columns
for i in (np.arange(0,len(cnames))):
    xvars = list(cnames)
    yvar = xvars.pop(i)
    mod = sm.OLS(x_train[yvar],sm.add_constant(x_train_cons[xvars]))
    res = mod.fit()
    vif = 1/(1-res.rsquared)
    print(yvar,round(vif,3))

Iteration-2                             OLS Regression Results                            
Dep. Variable:                   MEDV   R-squared:                       0.738
Model:                            OLS   Adj. R-squared:                  0.729
Method:                 Least Squares   F-statistic:                     80.14
Date:                Mon, 12 Apr 2021   Prob (F-statistic):           1.65e-91
Time:                        07:48:29   Log-Likelihood:                -1038.5
No. Observations:                 354   AIC:                             2103.
Df Residuals:                     341   BIC:                             2153.
Df Model:                          12                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         38.3247      5.932      6.

#### In the 2nd Iteration, R-Square Value is 0.738 and Adjacent R-Square Value is 0.729. There is not that much difference.
#### Executing one more iteration by removing INDUS variable.

In [14]:
x_columns = ["CRIM","ZN","CHAS","NOX","RM","DIS","RAD","TAX","PTRATIO","B","LSTAT"]

pdx = housing_data[x_columns]
pdy = housing_data["MEDV"]

x_train, x_test,y_train,y_test = train_test_split(pdx,pdy,train_size=70/100,random_state = 70)

x_train_cons = sm.add_constant (x_train)
x_test_cons = sm.add_constant (x_test)
full_mod = sm.OLS(y_train,x_train_cons)

regression_full = full_mod.fit()
print("Iteration-3", regression_full.summary())

print("Variance Inflation Factor")
cnames = x_train.columns
for i in (np.arange(0,len(cnames))):
    xvars = list(cnames)
    yvar = xvars.pop(i)
    mod = sm.OLS(x_train[yvar],sm.add_constant(x_train_cons[xvars]))
    res = mod.fit()
    vif = 1/(1-res.rsquared)
    print(yvar,round(vif,3))

Iteration-3                             OLS Regression Results                            
Dep. Variable:                   MEDV   R-squared:                       0.738
Model:                            OLS   Adj. R-squared:                  0.729
Method:                 Least Squares   F-statistic:                     87.50
Date:                Mon, 12 Apr 2021   Prob (F-statistic):           2.24e-92
Time:                        07:48:30   Log-Likelihood:                -1038.8
No. Observations:                 354   AIC:                             2102.
Df Residuals:                     342   BIC:                             2148.
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         38.4584      5.925      6.

#### In the 3rd iteration too, there is not that much difference. R-Square Value is 0.738 and Adjacent R-Square Value is 0.729

In [15]:
y_pred = regression_full.predict(x_test_cons)
y_pred_df = pd.DataFrame(y_pred)
y_pred_df.columns= ['y_pred']

pred_data = pd.DataFrame(y_pred_df['y_pred'])
y_test_cons = pd.DataFrame(y_test)
y_test_cons.reset_index(inplace=True)
pred_data['y_test'] = pd.DataFrame(y_test_cons['MEDV'])

In [16]:
rsqd = r2_score(y_test_cons['MEDV'].tolist(),y_pred_df['y_pred'].tolist())
print('Test R-Squared Value:',round(rsqd,4))

Test R-Squared Value: 0.7356
